fuel-poa-coordinator/src/service.rs

```diff
@@ -0,0 +1,316 @@
+            .await
+    }
+
+    async fn produce_block(&mut self) -> anyhow::Result<()> {
+        // Ask the block producer to create the block
+        let block = self.signal_produce_block().await?;
+
+        // Send the block back to the txpool
+        // TODO: this probably must be done differently with multi-node configuration
+        self.import_block_events_tx
+            .send(ImportBlockBroadcast::PendingFuelBlockImported {
+                block: Arc::new(block),
+            })
+            .expect("Failed to import the generated block");
+
+        // Update last block time
+        self.last_block_created = Instant::now();
+
+        // Set timer for the next block
+        match self.trigger {
+            Trigger::Never => {
+                unreachable!("This mode will never produce blocks");
+            }
+            Trigger::Instant => {}
+            Trigger::Interval { block_time } => {
+                // TODO: instead of sleeping for `block_time`, subtract the time we used for processing
+                self.timer.set_timeout(block_time, OnConflict::Min).await;
+            }
+            Trigger::Hybrid {
+                max_block_time,
+                min_block_time,
+                max_tx_idle_time,
+            } => {
+                self.timer
+                    .set_timeout(max_block_time, OnConflict::Min)
+                    .await;
+
+                let consumable_gas = self.txpool.total_consumable_gas().await?;
+
+                // If txpool still has more than a full block of transactions available,
+                // produce new block in min_block_time.
+                if consumable_gas > self.block_gas_limit {
+                    self.timer
+                        .set_timeout(min_block_time, OnConflict::Min)
+                        .await;
+                } else if consumable_gas > 0 {
+                    // If we still have available txs, reduce the timeout to max idle time
+                    self.timer
+                        .set_timeout(max_tx_idle_time, OnConflict::Min)
+                        .await;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn on_txpool_event(
+        &mut self,
+        txpool_event: &TxStatusBroadcast,
+    ) -> anyhow::Result<()> {
+        match txpool_event.status {
+            TxStatus::Submitted => match self.trigger {
+                Trigger::Instant => {
+                    self.produce_block().await?;
+                    Ok(())
+                }
+                Trigger::Never | Trigger::Interval { .. } => Ok(()),
+                Trigger::Hybrid {
+                    max_tx_idle_time,
+                    min_block_time,
+                    ..
+                } => {
+                    let consumable_gas = self.txpool.total_consumable_gas().await?;
+
+                    // We have at least one transaction, so tx_max_idle_time is the limit
+                    self.timer
```

Comment:
[nit] Should we move this to an else clause below and only update the timer if we don't immediately jump into block production?

Reply:
Makes sense